File: 08_react_app/demo_video/record_segmentA.py
"""
Segment A recorder — automated FOI walkthrough (I drive), headless, records .webm.

Final-video order (user's narrative): Command Centre -> Cases -> Knowledge Base.
Unlike the rejected URL-flythrough, this performs REAL interactions:
  - word-cloud hover + click -> filtered cases drill-down
  - Cases focus lanes, hover precedent/complexity chips, switch to List view
  - open FOI-2026-0115: AI triage, precedent card, hash-chained audit trail
  - Knowledge Base semantic search ("personal data") -> cross-authority precedent

Runs against localhost:3000 (no SSO). Output: demo_video/raw/segmentA.webm
"""
import asyncio, os, re, shutil
from playwright.async_api import async_playwright, Page

BASE = "http://localhost:3000"
HERE = os.path.dirname(os.path.abspath(__file__))
RAW = os.path.join(HERE, "raw")
VIEWPORT = {"width": 1440, "height": 900}
WC_TITLE = re.compile(r"click to see these cases", re.I)


async def hold(page: Page, s: float):
    await page.wait_for_timeout(int(s * 1000))


async def slow_scroll(page: Page, step: int = 300, pause: int = 700, back: bool = True):
    total = await page.evaluate("() => document.body.scrollHeight")
    y = 0
    while y < total - VIEWPORT["height"]:
        y += step
        await page.evaluate(f"window.scrollTo({{top:{y},behavior:'smooth'}})")
        await page.wait_for_timeout(pause)
    await page.wait_for_timeout(1000)
    if back:
        await page.evaluate("window.scrollTo({top:0,behavior:'smooth'})")
        await page.wait_for_timeout(800)


async def try_hover(page: Page, locator, label: str, dwell: float = 2.0):
    try:
        el = locator.first
        await el.scroll_into_view_if_needed(timeout=5000)
        await el.hover(timeout=5000)
        await hold(page, dwell)
        print(f"  hover ok: {label}", flush=True)
    except Exception as e:  # noqa: BLE001
        print(f"  hover miss {label}: {e}", flush=True)


async def section_command_centre(page: Page):
    print("S1 Command Centre", flush=True)
    await page.goto(BASE + "/", wait_until="networkidle")
    # wait for clickable word-cloud terms to render
    try:
        await page.wait_for_function(
            "() => [...document.querySelectorAll('svg text')].some(t => /click to see these cases/i.test((t.querySelector('title')||{}).textContent||''))",
            timeout=20000,
        )
    except Exception:
        pass
    await hold(page, 3)
    # slow read of scorecard / SLA gauge / peer benchmark, then down to Cortex section
    await slow_scroll(page, step=300, pause=750, back=False)
    # find the top word-cloud term, hover then click -> drill-down
    term = await page.evaluate(
        "() => { const n=[...document.querySelectorAll('svg text')].filter(t=>/click to see these cases/i.test((t.querySelector('title')||{}).textContent||'')); if(!n.length) return null; const t=n[0].querySelector('title').textContent.match(/^(.+?):/); return t?t[1]:null }"
    )
    if term:
        wc = page.locator("svg text", has_text=term).first
        await try_hover(page, wc, f"word-cloud '{term}'", dwell=2.0)
        try:
            await wc.click(timeout=5000)
            await page.wait_for_url(re.compile(r"/cases\?.*keyword="), timeout=8000)
            await page.wait_for_timeout(500)
            try:
                await page.get_by_text(re.compile(r"Showing cases mentioning", re.I)).first.wait_for(timeout=6000)
            except Exception:
                pass
            await hold(page, 4)
            print(f"  drilldown -> {page.url}", flush=True)
        except Exception as e:  # noqa: BLE001
            print(f"  word-cloud click miss: {e}", flush=True)


async def section_cases(page: Page):
    print("S2 Cases", flush=True)
    await page.goto(BASE + "/cases", wait_until="networkidle")
    try:
        await page.get_by_text("Quick wins", exact=False).first.wait_for(timeout=12000)
    except Exception:
        pass
    await hold(page, 4)
    # hover a precedent pill and a complexity chip (tooltips)
    await try_hover(page, page.get_by_text(re.compile(r"% precedent match", re.I)), "precedent pill", 2.5)
    await try_hover(page, page.locator("[class*='complexity'], [aria-label*='omplexity']"), "complexity chip", 2.5)
    # switch to List view if a toggle exists
    for name in ["List", "List view", "Table"]:
        try:
            await page.get_by_role("button", name=re.compile(name, re.I)).first.click(timeout=3000)
            await hold(page, 3)
            print(f"  switched to {name} view", flush=True)
            break
        except Exception:
            continue
    await slow_scroll(page, step=320, pause=650, back=False)


async def section_case_detail(page: Page):
    print("S2 Case detail FOI-2026-0115", flush=True)
    await page.goto(BASE + "/cases/FOI-2026-0115", wait_until="networkidle")
    try:
        await page.get_by_text(re.compile(r"How AI triaged this case|AI triage", re.I)).first.wait_for(timeout=12000)
    except Exception:
        pass
    await hold(page, 4)
    await slow_scroll(page, step=280, pause=750, back=False)
    # ensure the audit trail (Chain verified) gets screen time
    try:
        await page.get_by_text(re.compile(r"AI evidence & audit trail", re.I)).first.scroll_into_view_if_needed(timeout=6000)
        await hold(page, 4)
    except Exception:
        pass


async def section_knowledge(page: Page):
    print("S4 Knowledge Base", flush=True)
    await page.goto(BASE + "/guidance", wait_until="networkidle")
    await hold(page, 3)
    await slow_scroll(page, step=320, pause=650, back=True)
    try:
        box = page.get_by_placeholder(re.compile(r"search", re.I)).first
        await box.scroll_into_view_if_needed(timeout=6000)
        await box.click()
        await hold(page, 0.6)
        await box.fill("personal data")
        await hold(page, 0.6)
        await box.press("Enter")
        await page.wait_for_timeout(2800)
        await hold(page, 3)
        await slow_scroll(page, step=300, pause=700, back=False)
        print("  KB search done", flush=True)
    except Exception as e:  # noqa: BLE001
        print(f"  KB search miss: {e}", flush=True)


async def main():
    os.makedirs(RAW, exist_ok=True)
    async with async_playwright() as p:
        browser = await p.chromium.launch(headless=True)
        ctx = await browser.new_context(viewport=VIEWPORT, record_video_dir=RAW, record_video_size=VIEWPORT)
        page = await ctx.new_page()
        await section_command_centre(page)
        await section_cases(page)
        await section_case_detail(page)
        await section_knowledge(page)
        video = page.video
        await ctx.close()
        out = None
        if video:
            src = await video.path()
            out = os.path.join(RAW, "segmentA.webm")
            shutil.copyfile(src, out)
            print(f"SEGMENT_A_SAVED {out}", flush=True)
        await browser.close()
        print("DONE", flush=True)


if __name__ == "__main__":
    asyncio.run(main())

File: 08_react_app/demo_video/record_segmentA.mjs
// Segment A recorder — automated FOI walkthrough (I drive), headless, records .webm.
// Order (user's narrative): Command Centre -> Cases -> Knowledge Base, with REAL interactions.
// localhost:3000 (no SSO). Output: demo_video/raw/segmentA.webm
import { chromium } from "playwright"
import fs from "node:fs"
import path from "node:path"

const BASE = process.env.BASE || "http://localhost:3000"
const RAW = path.resolve("demo_video/raw")
const VP = { width: 1440, height: 900 }
fs.mkdirSync(RAW, { recursive: true })
const log = (m) => console.log(m)
const hold = (page, s) => page.waitForTimeout(Math.round(s * 1000))

async function slowScroll(page, { step = 300, pause = 700, back = true } = {}) {
  const total = await page.evaluate(() => document.body.scrollHeight)
  let y = 0
  while (y < total - VP.height) {
    y += step
    await page.evaluate((yy) => window.scrollTo({ top: yy, behavior: "smooth" }), y)
    await page.waitForTimeout(pause)
  }
  await page.waitForTimeout(1000)
  if (back) { await page.evaluate(() => window.scrollTo({ top: 0, behavior: "smooth" })); await page.waitForTimeout(800) }
}

async function tryHover(page, locator, label, dwell = 2.0) {
  try {
    const el = locator.first()
    await el.scrollIntoViewIfNeeded({ timeout: 5000 })
    await el.hover({ timeout: 5000 })
    await hold(page, dwell)
    log(`  hover ok: ${label}`)
  } catch (e) { log(`  hover miss ${label}: ${e.message}`) }
}

async function commandCentre(page) {
  log("S1 Command Centre")
  await page.goto(`${BASE}/`, { waitUntil: "networkidle" })
  await page.waitForFunction(
    () => [...document.querySelectorAll("svg text")].some(t => /click to see these cases/i.test((t.querySelector("title") || {}).textContent || "")),
    { timeout: 20000 }
  ).catch(() => {})
  await hold(page, 3)
  await slowScroll(page, { step: 300, pause: 750, back: false })
  const term = await page.evaluate(() => {
    const n = [...document.querySelectorAll("svg text")].filter(t => /click to see these cases/i.test((t.querySelector("title") || {}).textContent || ""))
    if (!n.length) return null
    const m = n[0].querySelector("title").textContent.match(/^(.+?):/)
    return m ? m[1] : null
  })
  if (term) {
    const wc = page.locator("svg text", { hasText: term }).first()
    await tryHover(page, wc, `word-cloud '${term}'`, 2.0)
    try {
      await wc.click({ timeout: 5000 })
      await page.waitForURL(/\/cases\?.*keyword=/, { timeout: 8000 })
      await page.getByText(/Showing cases mentioning/i).first().waitFor({ timeout: 6000 }).catch(() => {})
      await hold(page, 4)
      log(`  drilldown -> ${page.url()}`)
    } catch (e) { log(`  word-cloud click miss: ${e.message}`) }
  }
}

async function cases(page) {
  log("S2 Cases")
  await page.goto(`${BASE}/cases`, { waitUntil: "networkidle" })
  await page.getByText("Quick wins", { exact: false }).first().waitFor({ timeout: 12000 }).catch(() => {})
  await hold(page, 4)
  await tryHover(page, page.getByText(/% precedent match/i), "precedent pill", 2.5)
  await tryHover(page, page.locator("[class*='complexity'],[aria-label*='omplexity']"), "complexity chip", 2.5)
  for (const name of ["List view", "List", "Table"]) {
    try { await page.getByRole("button", { name: new RegExp(name, "i") }).first().click({ timeout: 3000 }); await hold(page, 3); log(`  switched to ${name}`); break } catch { /* next */ }
  }
  await slowScroll(page, { step: 320, pause: 650, back: false })
}

async function caseDetail(page) {
  log("S2 Case detail FOI-2026-0115")
  await page.goto(`${BASE}/cases/FOI-2026-0115`, { waitUntil: "networkidle" })
  await page.getByText(/How AI triaged this case|AI triage/i).first().waitFor({ timeout: 12000 }).catch(() => {})
  await hold(page, 4)
  await slowScroll(page, { step: 280, pause: 750, back: false })
  try {
    await page.getByText(/AI evidence & audit trail/i).first().scrollIntoViewIfNeeded({ timeout: 6000 })
    await hold(page, 4)
  } catch { /* ok */ }
}

async function knowledge(page) {
  log("S4 Knowledge Base")
  await page.goto(`${BASE}/guidance`, { waitUntil: "networkidle" })
  await hold(page, 3)
  await slowScroll(page, { step: 320, pause: 650, back: true })
  try {
    const box = page.getByPlaceholder(/search/i).first()
    await box.scrollIntoViewIfNeeded({ timeout: 6000 })
    await box.click(); await hold(page, 0.6)
    await box.fill("personal data"); await hold(page, 0.6)
    await box.press("Enter")
    await page.waitForTimeout(2800)
    await hold(page, 3)
    await slowScroll(page, { step: 300, pause: 700, back: false })
    log("  KB search done")
  } catch (e) { log(`  KB search miss: ${e.message}`) }
}

async function main() {
  const browser = await chromium.launch({ headless: true })
  const ctx = await browser.newContext({ viewport: VP, recordVideo: { dir: RAW, size: VP } })
  const page = await ctx.newPage()
  await commandCentre(page)
  await cases(page)
  await caseDetail(page)
  await knowledge(page)
  const video = page.video()
  await ctx.close()
  if (video) {
    const src = await video.path()
    const out = path.join(RAW, "segmentA.webm")
    fs.copyFileSync(src, out)
    log(`SEGMENT_A_SAVED ${out}`)
  }
  await browser.close()
  log("DONE")
}
main().catch((e) => { console.error("RECORD ERROR", e); process.exit(1) })
